File: python/src/vcp/negotiation.py
"""VCP 3.1 capability negotiation.

Implements the canonical ``vcp-hello`` / ``vcp-ack`` handshake defined by
the VCP-Spec capability-negotiation document and exercised by
``conformance/extensions/capability_negotiation.json``.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from typing import Any, TypeGuard

from .identity.token import Token

SEMVER_MINOR_PATTERN = re.compile(r"^(?:0|[1-9][0-9]{0,8})\.(?:0|[1-9][0-9]{0,8})$")
EXTENSION_PATTERN = re.compile(r"^VCP-X-[A-Za-z][A-Za-z0-9-]*$")
MAX_HANDSHAKE_BYTES = 65_536
MAX_EXTENSION_COUNT = 256
MAX_EXTENSION_LENGTH = 128
MAX_SUPPORTED_VERSIONS = 64
CORE_FEATURE_NAMES = (
    "encryption",
    "injection_scanning",
    "revocation",
    "audit_chain",
    "context_opacity",
)


def _snapshot(value: Any, label: str) -> Any:
    """Take a bounded JSON snapshot so caller mutation cannot alter a result."""
    try:
        encoded = json.dumps(value, ensure_ascii=False, allow_nan=False, separators=(",", ":"))
    except (TypeError, ValueError) as exc:
        raise TypeError(f"{label} must be JSON serializable") from exc
    if len(encoded.encode("utf-8")) > MAX_HANDSHAKE_BYTES:
        raise ValueError(f"{label} exceeds the 64 KiB wire limit")
    return json.loads(encoded)


def _version_parts(value: Any, field_name: str) -> tuple[str, str]:
    if not isinstance(value, str) or SEMVER_MINOR_PATTERN.fullmatch(value) is None:
        raise ValueError(f"{field_name} must be a semver major.minor string")
    major, minor = value.split(".", 1)
    return major, minor


def _is_minor_version(value: Any) -> TypeGuard[str]:
    return isinstance(value, str) and SEMVER_MINOR_PATTERN.fullmatch(value) is not None


def _numeric_text_key(component: str) -> tuple[int, str]:
    normalized = component.lstrip("0") or "0"
    return len(normalized), normalized


def _version_key(value: str) -> tuple[tuple[int, str], tuple[int, str]]:
    major, minor = _version_parts(value, "version")
    return _numeric_text_key(major), _numeric_text_key(minor)


def _validate_core_features(value: Any) -> dict[str, bool]:
    if not isinstance(value, dict):
        raise ValueError("server.core_features must be an object")
    result: dict[str, bool] = {}
    for name in CORE_FEATURE_NAMES:
        feature = value.get(name)
        if not isinstance(feature, bool):
            raise ValueError(f"server.core_features.{name} must be a boolean")
        result[name] = feature
    for name, feature in value.items():
        if name not in result:
            if not isinstance(name, str) or not isinstance(feature, bool):
                raise ValueError("additional core feature entries must map strings to booleans")
            result[name] = feature
    return result


def _requested_extensions(value: Any) -> list[str]:
    if not isinstance(value, list) or len(value) > MAX_EXTENSION_COUNT:
        raise ValueError("extensions must be a bounded array")
    requested: list[str] = []
    seen: set[str] = set()
    for extension in value:
        if not isinstance(extension, str) or not extension or len(extension) > MAX_EXTENSION_LENGTH:
            raise ValueError("extension requests must be strings of 1 to 128 characters")
        if extension in seen:
            raise ValueError("extension requests must be unique")
        seen.add(extension)
        # Section 3.1 requires malformed identifiers to be ignored.
        if EXTENSION_PATTERN.fullmatch(extension) is None:
            continue
        requested.append(extension)
    return requested


@dataclass
class VCPHello:
    """Canonical client ``vcp-hello`` message."""

    version: str
    extensions: list[str] = field(default_factory=list)
    identity: str | None = None
    min_version: str = "1.0"
    client_id: str | None = None
    type: str = field(default="vcp-hello", init=False)

    def __post_init__(self) -> None:
        _version_parts(self.version, "version")
        _version_parts(self.min_version, "min_version")
        if _version_key(self.min_version) > _version_key(self.version):
            raise ValueError("min_version must not exceed version")
        self.extensions = _requested_extensions(self.extensions)
        if self.identity is not None and (
            not isinstance(self.identity, str) or len(self.identity) > 2048
        ):
            raise ValueError("identity must be a string of at most 2048 characters or null")
        if self.client_id is not None and (
            not isinstance(self.client_id, str) or not self.client_id or len(self.client_id) > 256
        ):
            raise ValueError("client_id must be a string of 1 to 256 characters")

    def to_dict(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "type": self.type,
            "version": self.version,
            "extensions": list(self.extensions),
            "identity": self.identity,
            "min_version": self.min_version,
        }
        if self.client_id is not None:
            result["client_id"] = self.client_id
        return result

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> VCPHello:
        if not isinstance(data, dict) or data.get("type") != "vcp-hello":
            raise ValueError("VCP-Hello must be an object with type 'vcp-hello'")
        return cls(
            version=data["version"],
            extensions=data.get("extensions", []),
            identity=data.get("identity"),
            min_version=data.get("min_version", "1.0"),
            client_id=data.get("client_id"),
        )


@dataclass
class VCPAck:
    """Canonical successful ``vcp-ack`` response."""

    version: str
    supported: list[str]
    unsupported: list[str]
    capabilities: dict[str, dict[str, Any]]
    core_features: dict[str, bool]
    server_id: str | None = None
    session_id: str | None = None
    type: str = field(default="vcp-ack", init=False)

    def __post_init__(self) -> None:
        _version_parts(self.version, "version")
        self.supported = _requested_extensions(self.supported)
        self.unsupported = _requested_extensions(self.unsupported)
        if set(self.supported) & set(self.unsupported):
            raise ValueError("supported and unsupported extensions must be disjoint")
        if not isinstance(self.capabilities, dict) or any(
            key not in self.supported or not isinstance(value, dict)
            for key, value in self.capabilities.items()
        ):
            raise ValueError("capabilities must contain objects only for supported extensions")
        self.capabilities = _snapshot(self.capabilities, "capabilities")
        self.core_features = _validate_core_features(self.core_features)
        for field_name, value, maximum in (
            ("server_id", self.server_id, 256),
            ("session_id", self.session_id, 128),
        ):
            if value is not None and (
                not isinstance(value, str) or not value or len(value) > maximum
            ):
                raise ValueError(f"{field_name} must be a string of 1 to {maximum} characters")

    @property
    def active_extensions(self) -> list[str]:
        """Deprecated alias for callers migrating from the pre-spec API."""
        return self.supported

    @property
    def rejected_extensions(self) -> list[str]:
        """Deprecated alias for callers migrating from the pre-spec API."""
        return self.unsupported

    def to_dict(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "type": self.type,
            "version": self.version,
            "supported": list(self.supported),
            "unsupported": list(self.unsupported),
            "capabilities": _snapshot(self.capabilities, "capabilities"),
            "core_features": dict(self.core_features),
        }
        if self.server_id is not None:
            result["server_id"] = self.server_id
        if self.session_id is not None:
            result["session_id"] = self.session_id
        return result

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> VCPAck:
        if not isinstance(data, dict) or data.get("type") != "vcp-ack":
            raise ValueError("VCP-Ack must be an object with type 'vcp-ack'")
        return cls(
            version=data["version"],
            supported=data["supported"],
            unsupported=data["unsupported"],
            capabilities=data["capabilities"],
            core_features=data["core_features"],
            server_id=data.get("server_id"),
            session_id=data.get("session_id"),
        )


@dataclass
class VCPError:
    """Structured ``vcp-error`` handshake response."""

    code: str
    message: str
    supported_versions: list[str] | None = None
    retry_after: int | None = None
    type: str = field(default="vcp-error", init=False)

    def to_dict(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "type": self.type,
            "code": self.code,
            "message": self.message,
            "retry_after": self.retry_after,
        }
        if self.supported_versions is not None:
            result["supported_versions"] = list(self.supported_versions)
        return result


def _server_versions(server: dict[str, Any]) -> list[str]:
    values = server.get("supported_versions")
    if not isinstance(values, list) or not values or len(values) > MAX_SUPPORTED_VERSIONS:
        raise ValueError(
            "server.supported_versions must be a non-empty array of at most 64 entries"
        )
    versions: set[str] = set()
    for value in values:
        _version_parts(value, "server supported version")
        versions.add(value)
    return sorted(versions, key=_version_key)


def negotiate_handshake(client_hello: dict[str, Any], server: dict[str, Any]) -> dict[str, Any]:
    """Negotiate a canonical wire-level capability handshake."""
    hello_data = _snapshot(client_hello, "VCP-Hello")
    server_data = _snapshot(server, "server capability configuration")
    if not isinstance(hello_data, dict) or hello_data.get("type") != "vcp-hello":
        raise ValueError("VCP-Hello must be an object with type 'vcp-hello'")
    if not isinstance(server_data, dict):
        raise ValueError("server capability configuration must be an object")

    server_versions = _server_versions(server_data)
    client_version = hello_data.get("version")
    min_version = hello_data.get("min_version", "1.0")
    # A malformed client version range is a version mismatch, not a transport
    # fault: answer with the canonical VERSION_UNSUPPORTED error (Rust/WebMCP parity).
    if (
        not _is_minor_version(client_version)
        or not _is_minor_version(min_version)
        or _version_key(min_version) > _version_key(client_version)
    ):
        return VCPError(
            code="VERSION_UNSUPPORTED",
            message="No mutually supported VCP version",
            supported_versions=server_versions,
        ).to_dict()

    requested = _requested_extensions(hello_data.get("extensions", []))
    client_id = hello_data.get("client_id")
    if client_id is not None and (
        not isinstance(client_id, str) or not client_id or len(client_id) > 256
    ):
        raise ValueError("client_id must be a string of 1 to 256 characters")
    identity = hello_data.get("identity")
    if identity is not None:
        if not isinstance(identity, str) or len(identity) > 2048:
            raise ValueError("identity must be a string or null")
        try:
            Token.parse(identity)
        except (TypeError, ValueError):
            return VCPError(
                code="IDENTITY_INVALID",
                message="The supplied VCP/I identity token is invalid",
            ).to_dict()

    candidates = [
        version
        for version in server_versions
        if _version_key(min_version) <= _version_key(version) <= _version_key(client_version)
    ]
    if not candidates:
        return VCPError(
            code="VERSION_UNSUPPORTED",
            message="No mutually supported VCP version",
            supported_versions=server_versions,
        ).to_dict()
    version = candidates[-1]

    advertised_extensions = server_data.get("extensions", {})
    if not isinstance(advertised_extensions, dict):
        raise ValueError("server.extensions must be an object")
    if len(advertised_extensions) > MAX_EXTENSION_COUNT:
        raise ValueError("server.extensions exceeds 256 entries")
    for extension, capabilities in advertised_extensions.items():
        if (
            not isinstance(extension, str)
            or len(extension) > MAX_EXTENSION_LENGTH
            or EXTENSION_PATTERN.fullmatch(extension) is None
            or not isinstance(capabilities, dict)
        ):
            raise ValueError("server extensions must map VCP-X-* identifiers to objects")

    extensions_available = _version_key(version) >= _version_key("3.1")
    supported = [
        extension
        for extension in requested
        if extensions_available and extension in advertised_extensions
    ]
    supported_set = set(supported)
    unsupported = [extension for extension in requested if extension not in supported_set]
    capabilities = {extension: advertised_extensions[extension] for extension in supported}
    if "VCP-X-Torch" in supported_set and "VCP-X-Relational" not in supported_set:
        capabilities["VCP-X-Torch"]["degraded"] = True
    if "VCP-X-Intent" in supported_set and "VCP-X-Personal" not in supported_set:
        capabilities["VCP-X-Intent"]["personal_signals"] = False
    ack = VCPAck(
        version=version,
        supported=supported,
        unsupported=unsupported,
        capabilities=capabilities,
        core_features=_validate_core_features(server_data.get("core_features")),
        server_id=server_data.get("server_id"),
        session_id=server_data.get("session_id"),
    )
    return ack.to_dict()


def negotiate(
    hello: VCPHello,
    server_capabilities: dict[str, bool | dict[str, Any]],
    *,
    supported_versions: list[str] | None = None,
    core_features: dict[str, bool] | None = None,
) -> VCPAck | VCPError:
    """Object-oriented wrapper around :func:`negotiate_handshake`."""
    if not isinstance(hello, VCPHello):
        raise TypeError("hello must be a VCPHello")
    if not isinstance(server_capabilities, dict):
        raise TypeError("server_capabilities must be an object")
    extensions: dict[str, dict[str, Any]] = {}
    for name, capability in server_capabilities.items():
        if capability is True:
            extensions[name] = {}
        elif isinstance(capability, dict):
            extensions[name] = capability
        elif capability is not False:
            raise ValueError("server capabilities must be booleans or capability objects")
    server = {
        "supported_versions": (
            ["1.0", "2.0", "3.0", "3.1"] if supported_versions is None else supported_versions
        ),
        "extensions": extensions,
        "core_features": (
            {name: False for name in CORE_FEATURE_NAMES} if core_features is None else core_features
        ),
    }
    result = negotiate_handshake(hello.to_dict(), server)
    if result["type"] == "vcp-error":
        return VCPError(
            code=result["code"],
            message=result["message"],
            supported_versions=result.get("supported_versions"),
            retry_after=result.get("retry_after"),
        )
    return VCPAck.from_dict(result)

File: webmcp/src/extensions/negotiation.ts
/**
 * VCP 3.1 capability negotiation.
 *
 * The public shapes intentionally mirror the normative wire names. Keeping
 * snake_case here makes it harder for an implementation-only camelCase model
 * to drift from the protocol again.
 */

export const VCPExtension = {
  PERSONAL: 'VCP-X-Personal',
  RELATIONAL: 'VCP-X-Relational',
  CONSENSUS: 'VCP-X-Consensus',
  TORCH: 'VCP-X-Torch',
  INTENT: 'VCP-X-Intent',
} as const;

export type VCPExtension = (typeof VCPExtension)[keyof typeof VCPExtension];

/** @deprecated Use VCPExtension. Retained as a source-compatible export name. */
export const VCPCapability = VCPExtension;
/** @deprecated Use VCPExtension. */
export type VCPCapability = VCPExtension;

export interface VCPHello {
  readonly type: 'vcp-hello';
  /** Highest VCP major.minor version the client supports. */
  readonly version: string;
  readonly extensions?: readonly string[];
  readonly identity?: string | null;
  /** Lowest acceptable VCP version. Defaults to 1.0. */
  readonly min_version?: string;
  readonly client_id?: string;
}

export interface VCPCoreFeatures {
  readonly encryption: boolean;
  readonly injection_scanning: boolean;
  readonly revocation: boolean;
  readonly audit_chain: boolean;
  readonly context_opacity: boolean;
}

export interface VCPServerCapabilities {
  readonly supported_versions: readonly string[];
  /** Map of extension identifier to its capability advertisement. */
  readonly extensions: Readonly<Record<string, Readonly<Record<string, unknown>>>>;
  readonly core_features: VCPCoreFeatures;
  readonly server_id?: string;
  readonly session_id?: string;
}

export interface VCPAck {
  readonly type: 'vcp-ack';
  readonly version: string;
  readonly supported: readonly string[];
  readonly unsupported: readonly string[];
  readonly capabilities: Readonly<Record<string, Readonly<Record<string, unknown>>>>;
  readonly core_features: VCPCoreFeatures;
  readonly server_id?: string;
  readonly session_id?: string;
}

export interface VCPError {
  readonly type: 'vcp-error';
  readonly code: 'VERSION_UNSUPPORTED';
  readonly message: string;
  readonly supported_versions: readonly string[];
  readonly retry_after: null;
}

export type VCPNegotiationResult = VCPAck | VCPError;

const VERSION_PATTERN = /^(0|[1-9][0-9]{0,8})\.(0|[1-9][0-9]{0,8})$/;
const EXTENSION_PATTERN = /^VCP-X-[A-Za-z][A-Za-z0-9-]*$/;
const MAX_REQUESTED_EXTENSIONS = 256;
const MAX_SERVER_VERSIONS = 64;
const MAX_EXTENSION_IDENTIFIER_LENGTH = 128;
const MAX_DIAGNOSTIC_IDENTIFIER_LENGTH = 256;
const MAX_SESSION_IDENTIFIER_LENGTH = 128;
const MAX_IDENTITY_LENGTH = 2048;
const MAX_HANDSHAKE_BYTES = 64 * 1024;

interface ParsedVersion {
  readonly source: string;
  readonly major: number;
  readonly minor: number;
}

function parseVersion(value: unknown): ParsedVersion | null {
  if (typeof value !== 'string') return null;
  const match = VERSION_PATTERN.exec(value);
  if (!match) return null;
  const major = Number(match[1]);
  const minor = Number(match[2]);
  if (!Number.isSafeInteger(major) || !Number.isSafeInteger(minor)) return null;
  return { source: value, major, minor };
}

function compareVersions(a: ParsedVersion, b: ParsedVersion): number {
  return a.major === b.major ? a.minor - b.minor : a.major - b.major;
}

function requireBoundedArray(value: unknown, name: string, limit: number): readonly unknown[] {
  if (!Array.isArray(value)) throw new TypeError(`${name} must be an array`);
  if (value.length > limit) throw new RangeError(`${name} exceeds ${limit} entries`);
  return value;
}

/** JSON Schema maxLength counts Unicode code points, not UTF-16 code units. */
function characterLength(value: string): number {
  return Array.from(value).length;
}

function validRequestedExtensions(value: unknown): string[] {
  const entries = value === undefined
    ? []
    : requireBoundedArray(value, 'hello.extensions', MAX_REQUESTED_EXTENSIONS);
  const seen = new Set<string>();
  const valid: string[] = [];
  for (const entry of entries) {
    if (typeof entry !== 'string') {
      throw new TypeError('hello.extensions entries must be strings');
    }
    const length = characterLength(entry);
    if (length === 0 || length > MAX_EXTENSION_IDENTIFIER_LENGTH) {
      throw new RangeError('hello.extensions entries must contain 1 to 128 characters');
    }
    if (seen.has(entry)) throw new TypeError(`duplicate hello extension: ${entry}`);
    seen.add(entry);
    // Bounded strings outside the extension namespace are ignored by spec.
    if (!EXTENSION_PATTERN.test(entry)) continue;
    valid.push(entry);
  }
  return valid;
}

function validateCoreFeatures(value: unknown): VCPCoreFeatures {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new TypeError('server.core_features must be an object');
  }
  const record = value as Record<string, unknown>;
  const names = [
    'encryption',
    'injection_scanning',
    'revocation',
    'audit_chain',
    'context_opacity',
  ] as const;
  for (const name of names) {
    if (typeof record[name] !== 'boolean') {
      throw new TypeError(`server.core_features.${name} must be a boolean`);
    }
  }
  return {
    encryption: record.encryption as boolean,
    injection_scanning: record.injection_scanning as boolean,
    revocation: record.revocation as boolean,
    audit_chain: record.audit_chain as boolean,
    context_opacity: record.context_opacity as boolean,
  };
}

function copyOptionalIdentifier(
  value: unknown,
  name: string,
  maximum = MAX_DIAGNOSTIC_IDENTIFIER_LENGTH,
): string | undefined {
  if (value === undefined) return undefined;
  if (
    typeof value !== 'string' ||
    characterLength(value) === 0 ||
    characterLength(value) > maximum
  ) {
    throw new TypeError(`${name} must be a non-empty string of at most ${maximum} characters`);
  }
  return value;
}

function copyCapability(value: Readonly<Record<string, unknown>>, name: string): Record<string, unknown> {
  try {
    const encoded = JSON.stringify(value, (_key, candidate: unknown) => {
      if (
        candidate === undefined ||
        typeof candidate === 'function' ||
        typeof candidate === 'symbol' ||
        typeof candidate === 'bigint' ||
        (typeof candidate === 'number' && !Number.isFinite(candidate))
      ) {
        throw new TypeError('not a JSON value');
      }
      return candidate;
    });
    if (encoded === undefined) throw new TypeError('not JSON serializable');
    return JSON.parse(encoded) as Record<string, unknown>;
  } catch {
    throw new TypeError(`${name} must be a JSON-serializable object`);
  }
}

function assertHandshakeSize(value: unknown): void {
  let encoded: string;
  try {
    const result = JSON.stringify(value, (_key, candidate: unknown) => {
      if (typeof candidate === 'number' && !Number.isFinite(candidate)) {
        throw new TypeError('not a finite JSON number');
      }
      return candidate;
    });
    if (result === undefined) throw new TypeError('not JSON serializable');
    encoded = result;
  } catch {
    throw new TypeError('handshake must be JSON serializable');
  }
  if (new TextEncoder().encode(encoded).byteLength > MAX_HANDSHAKE_BYTES) {
    throw new RangeError('handshake exceeds 64 KiB');
  }
}

function versionError(serverVersions: readonly ParsedVersion[]): VCPError {
  return {
    type: 'vcp-error',
    code: 'VERSION_UNSUPPORTED',
    message: 'No mutually supported VCP version',
    supported_versions: serverVersions.map((version) => version.source),
    retry_after: null,
  };
}

/**
 * Negotiate the highest mutually supported VCP version and requested extensions.
 * Invalid extension identifiers are ignored. Duplicate raw request strings are
 * rejected before filtering. Extensions are negotiated only at 3.1+.
 */
export function negotiate(
  hello: VCPHello,
  server: VCPServerCapabilities,
): VCPNegotiationResult {
  if (typeof hello !== 'object' || hello === null || hello.type !== 'vcp-hello') {
    throw new TypeError("hello.type must be 'vcp-hello'");
  }
  if (typeof server !== 'object' || server === null) {
    throw new TypeError('server must be an object');
  }
  assertHandshakeSize(hello);

  copyOptionalIdentifier(hello.client_id, 'hello.client_id');
  if (
    hello.identity !== undefined &&
    hello.identity !== null &&
    (typeof hello.identity !== 'string' ||
      characterLength(hello.identity) > MAX_IDENTITY_LENGTH)
  ) {
    throw new TypeError('hello.identity must be null or a string of at most 2048 characters');
  }

  const serverEntries = requireBoundedArray(
    server.supported_versions,
    'server.supported_versions',
    MAX_SERVER_VERSIONS,
  );
  const byVersion = new Map<string, ParsedVersion>();
  for (const entry of serverEntries) {
    const parsed = parseVersion(entry);
    if (!parsed) throw new TypeError('server.supported_versions entries must be major.minor');
    byVersion.set(parsed.source, parsed);
  }
  const serverVersions = [...byVersion.values()].sort(compareVersions);
  if (serverVersions.length === 0) {
    throw new RangeError('server.supported_versions must be non-empty');
  }
  const clientMax = parseVersion(hello.version);
  const clientMin = parseVersion(hello.min_version ?? '1.0');
  if (!clientMax || !clientMin || compareVersions(clientMin, clientMax) > 0) {
    return versionError(serverVersions);
  }

  const negotiated = serverVersions
    .filter(
      (version) =>
        compareVersions(version, clientMin) >= 0 &&
        compareVersions(version, clientMax) <= 0,
    )
    .at(-1);
  if (!negotiated) return versionError(serverVersions);

  const requested = validRequestedExtensions(hello.extensions);
  if (
    typeof server.extensions !== 'object' ||
    server.extensions === null ||
    Array.isArray(server.extensions)
  ) {
    throw new TypeError('server.extensions must be an object');
  }
  if (Object.keys(server.extensions).length > MAX_REQUESTED_EXTENSIONS) {
    throw new RangeError('server.extensions exceeds 256 entries');
  }
  for (const [identifier, capability] of Object.entries(server.extensions)) {
    if (
      characterLength(identifier) > MAX_EXTENSION_IDENTIFIER_LENGTH ||
      !EXTENSION_PATTERN.test(identifier)
    ) {
      throw new TypeError(`server.extensions has invalid identifier: ${identifier}`);
    }
    if (typeof capability !== 'object' || capability === null || Array.isArray(capability)) {
      throw new TypeError(`server.extensions.${identifier} must be an object`);
    }
  }
  const coreFeatures = validateCoreFeatures(server.core_features);
  const extensionNegotiationAvailable = compareVersions(
    negotiated,
    { source: '3.1', major: 3, minor: 1 },
  ) >= 0;
  const supported = extensionNegotiationAvailable
    ? requested.filter((identifier) => Object.hasOwn(server.extensions, identifier))
    : [];
  const supportedSet = new Set(supported);
  const unsupported = requested.filter((identifier) => !supportedSet.has(identifier));
  const capabilities: Record<string, Readonly<Record<string, unknown>>> = {};
  for (const identifier of supported) {
    const capability = server.extensions[identifier];
    const snapshot = copyCapability(capability, `server.extensions.${identifier}`);
    if (identifier === VCPExtension.TORCH && !supportedSet.has(VCPExtension.RELATIONAL)) {
      snapshot.degraded = true;
    }
    if (identifier === VCPExtension.INTENT && !supportedSet.has(VCPExtension.PERSONAL)) {
      snapshot.personal_signals = false;
    }
    capabilities[identifier] = Object.freeze(snapshot);
  }

  const serverId = copyOptionalIdentifier(server.server_id, 'server.server_id');
  const sessionId = copyOptionalIdentifier(
    server.session_id,
    'server.session_id',
    MAX_SESSION_IDENTIFIER_LENGTH,
  );
  const result: VCPAck = {
    type: 'vcp-ack',
    version: negotiated.source,
    supported,
    unsupported,
    capabilities,
    core_features: coreFeatures,
    ...(serverId === undefined ? {} : { server_id: serverId }),
    ...(sessionId === undefined ? {} : { session_id: sessionId }),
  };
  assertHandshakeSize(result);
  return result;
}

/** Create a canonical VCP 3.1 hello requesting every defined extension. */
export function createFullHello(clientId?: string, identity?: string | null): VCPHello {
  return {
    type: 'vcp-hello',
    version: '3.1',
    min_version: '1.0',
    extensions: Object.values(VCPExtension),
    ...(identity === undefined ? {} : { identity }),
    ...(clientId === undefined ? {} : { client_id: clientId }),
  };
}
